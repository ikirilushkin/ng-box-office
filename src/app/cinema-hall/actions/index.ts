import * as CinemaHallActions from './cinema-hall.actions';

export { CinemaHallActions };
