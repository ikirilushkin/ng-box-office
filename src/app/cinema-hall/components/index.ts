export * from './order/order.component';
export * from './seat/seat.component';
