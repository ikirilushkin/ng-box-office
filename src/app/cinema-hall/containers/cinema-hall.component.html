<bo-seating-arrangement [hall]="hallView$ | async"></bo-seating-arrangement>
<bo-order [orderedSeats]="orderedSeats$ | async" (proceedOrder)="onOrderProceed($event)"
    (cancelOrder)="onCancelOrder($event)">
</bo-order>