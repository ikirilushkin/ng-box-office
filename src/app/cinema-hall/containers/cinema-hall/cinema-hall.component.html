<div class="cinema-hall">
    <bo-seating-arrangement></bo-seating-arrangement>
    <bo-order class="cinema-hall__order" [orderedSeats]="orderedSeats$ | async" (proceedOrder)="onOrderProceed($event)"
        (cancelOrder)="onCancelOrder($event)">
    </bo-order>
</div>