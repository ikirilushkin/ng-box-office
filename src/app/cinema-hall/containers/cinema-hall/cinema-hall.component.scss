.cinema-hall {
  display: flex;
}

.cinema-hall__order {
  margin-left: auto;
}
