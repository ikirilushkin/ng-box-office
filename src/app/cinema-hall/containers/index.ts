export * from './cinema-hall/cinema-hall.component';
export * from './seating-arrangement/seating-arrangement.component';
