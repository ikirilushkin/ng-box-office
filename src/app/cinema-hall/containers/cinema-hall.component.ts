import { Component, OnInit } from '@angular/core';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { cancelOrder, proceedOrder } from '../actions/cinema-hall.actions';
import { CinemaHallView, CinemaRow, Seat } from '../models';
import {
  getOrderedSeats,
  getSeatRows
} from '../selectors/cinema-hall.selectors';

@Component({
  selector: 'bo-cinema-hall',
  templateUrl: './cinema-hall.component.html',
  styleUrls: ['./cinema-hall.component.css']
})
export class CinemaHallComponent implements OnInit {
  hallView$: Observable<CinemaHallView>;

  orderedSeats$: Observable<Seat[]> = this.store.select(getOrderedSeats);

  constructor(private store: Store<{}>) {}

  ngOnInit() {
    this.hallView$ = this.store.select(getSeatRows).pipe(
      map(data => {
        const rows: CinemaRow[] = [];
        for (const key of Object.keys(data)) {
          const cinemaRow: CinemaRow = {
            rowNumber: +key,
            seats: data[key]
          };
          rows.push(cinemaRow);
        }
        return { rows };
      })
    );
  }

  onOrderProceed(ids: number[]) {
    this.store.dispatch(proceedOrder({ ids }));
  }

  onCancelOrder(ids: number[]) {
    this.store.dispatch(cancelOrder({ ids }));
  }
}
