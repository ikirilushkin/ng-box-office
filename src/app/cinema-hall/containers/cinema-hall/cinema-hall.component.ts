import { Component } from '@angular/core';
import { OrderActions } from '@bo/cinema-hall/actions';
import { Seat } from '@bo/cinema-hall/models';
import { getOrderedSeats } from '@bo/cinema-hall/selectors/cinema-hall.selectors';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';

@Component({
  selector: 'bo-cinema-hall',
  templateUrl: './cinema-hall.component.html',
  styleUrls: ['./cinema-hall.component.scss']
})
export class CinemaHallComponent {
  orderedSeats$: Observable<Seat[]> = this.store.select(getOrderedSeats);

  constructor(private store: Store<{}>) {}

  onOrderProceed(ids: number[]) {
    this.store.dispatch(OrderActions.proceedOrder({ ids }));
  }

  onCancelOrder(ids: number[]) {
    this.store.dispatch(OrderActions.cancelOrder({ ids }));
  }
}
