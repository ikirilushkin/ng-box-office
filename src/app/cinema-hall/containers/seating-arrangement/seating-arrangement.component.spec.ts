import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { SeatActions } from '@bo/cinema-hall/actions';
import { SeatComponent } from '@bo/cinema-hall/components';
import { SeatingArrangementComponent } from '@bo/cinema-hall/containers';
import { Seat, SeatStatus } from '@bo/cinema-hall/models';
import { getCinemaHallView } from '@bo/cinema-hall/selectors/cinema-hall.selectors';
import { Store } from '@ngrx/store';
import { MockStore, provideMockStore } from '@ngrx/store/testing';

describe('SeatingArrangementComponent', () => {
  let component: SeatingArrangementComponent;
  let fixture: ComponentFixture<SeatingArrangementComponent>;
  let store: MockStore<{}>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [SeatingArrangementComponent, SeatComponent],
      providers: [
        provideMockStore({
          selectors: [{ selector: getCinemaHallView, value: { rows: [] } }]
        })
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(SeatingArrangementComponent);
    component = fixture.componentInstance;
    store = TestBed.get(Store);
    spyOn(store, 'dispatch');
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a bookSeat event when free seat selected ', () => {
    const seat = generateSeat('FREE');
    const action = SeatActions.bookSeat({ seatId: seat.id });
    component.onToggleSeat(seat);
    expect(store.dispatch).toHaveBeenLastCalledWith(action);
  });

  it('should dispatch a cancelSeatBooking event when booked seat selected ', () => {
    const seat = generateSeat('BOOKED');
    const action = SeatActions.cancelSeatBooking({ seatId: seat.id });
    component.onToggleSeat(seat);
    expect(store.dispatch).toHaveBeenLastCalledWith(action);
  });
});

function generateSeat(status: SeatStatus): Seat {
  return {
    id: 1000,
    row: 1,
    seatNumber: 1,
    status
  };
}
