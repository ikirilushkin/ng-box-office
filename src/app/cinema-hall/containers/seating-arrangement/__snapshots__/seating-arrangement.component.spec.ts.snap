// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeatingArrangementComponent should create 1`] = `
<bo-seating-arrangement
  hall$={[Function Observable]}
  store={[Function MockStore]}
>
  
</bo-seating-arrangement>
`;
